File: 2017_May23/performance/ThreadPerformance.java
class Worker implements Runnable {
    Long y = 0L;
    public void run() {
        String w = Thread.currentThread().getName();

        System.out.println("Starting worker: " + w);
        for (Integer i=0; i<1000; i++) {
            for (Integer j=0; j<10000; j++) {
                y += i*j;
            }
        }
        System.out.println("Worker " + w + " complete");
    }
}

class ThreadPerformance {
    public static void startWorkers() throws Exception {
        Runnable worker = new Worker();
        Thread[] pool = new Thread[16];
        
        for (int i=0; i<16; i++) {
            pool[i] = new Thread(worker);
            pool[i].start();
        }
        System.out.println("Created 16 workers...");

        for (int i=0; i<16; i++) {
            pool[i].join();
        }
        System.out.println("All workers complete...");
    }

    public static void main(String[] args) throws Exception {
   
        long start = System.currentTimeMillis();
        startWorkers();
        long duration = System.currentTimeMillis() - start;
        System.out.printf("startWorkers: %f seconds.\n", duration / 1000.0);
    }
}

File: 2017_May23/performance/thread_performance.py
#from threading import Thread
from multiprocessing import Process as Thread

from performance import timeit, print_log

def worker(w):
    y = 0
    print "Starting worker:", w
    for i in range(1000000):
        for j in range(10000):
            y += i*j
    print "Worker", w, "complete..."

@timeit
def start_workers():
    pool = { }
    for i in range(16):
        pool[i] = Thread(target=worker, args=(i,))
        pool[i].start()
    print "Created 16 workers..."

    for i in range(16):
        pool[i].join()
    print "All workers complete..."

start_workers()

print_log()

